org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -299,20 +244,38 @@ public TypeBinding resolveType(BlockScope scope) {
 		this.pattern.setExpressionContext(ExpressionContext.INSTANCEOF_CONTEXT);
 		this.pattern.setExpectedType(this.expression.resolvedType);
 		this.pattern.resolveType(scope);
+
+		switch (this.expression.bits & ASTNode.RestrictiveFlagMASK) {
+
+			case Binding.FIELD:
+			case Binding.LOCAL:
+				// side effect free reevaluation possible, avoid wasting local slots.
+				break;
+			default:
+				// reevaluation may double jeopardize as side effects may recur, compute once and cache
+				LocalVariableBinding local =
+						new LocalVariableBinding(
+							InstanceOfExpression.SECRET_EXPRESSION_VALUE,
+							TypeBinding.wellKnownType(scope, T_JavaLangObject), // good enough, no need for sharper type.
+							ClassFileConstants.AccDefault,
+							false);
+				local.setConstant(Constant.NotAConstant);
+				local.useFlag = LocalVariableBinding.USED;
+				scope.addLocalVariable(local);
+				this.secretExpressionValue = local;
+				if (expressionType != TypeBinding.NULL)
+					this.secretExpressionValue.type = expressionType;
+		}
```

Comment:
We need to create a secret variable only if expression evaluation involves side effects. Otherwise, we can simply reevaluate the expression when needed rather than storing into and loading from a secret variable.